### PR TITLE
add factory for adw::PreferencePage

### PR DESCRIPTION
**Pull request by tronta**
#### Summary

My first try to add a factory

#### Checklist

- [x] cargo fmt
- [x] cargo clippy
- [x] cargo test
- [x] updated CHANGES.md

### CODE REVIEW

**Comment by tronta** [comment]
Hi @AaronErhardt , I tried a little to add this factory.
Maybe you can give some hints how to implement properly. Thanks

File: relm4/src/factory/widgets/libadwaita.rs [hunk 1]
```diff
@@ -56,6 +56,53 @@ impl FactoryView for adw::TabView {
...
+        self.remove(widget);
+    }
+
+    fn factory_append(
+        &self,
+        widget: impl AsRef<Self::Children>,
+        _position: &Self::Position,
+    ) -> Self::ReturnedWidget {
+        self.add(widget.as_ref());
+        *(&widget).as_ref()
```
Comment (AaronErhardt) [inline relm4/src/factory/widgets/libadwaita.rs:75]: Can't you simply use `widget.as_ref().clone()` at the end?

File: relm4/src/factory/widgets/libadwaita.rs [hunk 2]
```diff
@@ -56,6 +56,53 @@ impl FactoryView for adw::TabView {
...
+    }
+
+    fn factory_prepend(
+        &self,
+        widget: impl AsRef<Self::Children>,
+        _position: &(),
+    ) -> Self::ReturnedWidget {
+        self.add(widget.as_ref());
+        *(&widget).as_ref()
```
Comment (AaronErhardt) [inline relm4/src/factory/widgets/libadwaita.rs:84]: You can recycle `factory_append()` here, at least that saves one line of code.

**Review by AaronErhardt — CHANGES_REQUESTED** [review]
Looks good overall.

**Comment by tronta** [comment]
Cool thanks. One problem still remains… Maybe you have an other hint :smile:

**Comment by AaronErhardt** [comment]
> One problem still remains

What do you mean? Looks good to me.

File: relm4/src/factory/widgets/libadwaita.rs [hunk 3]
```diff
@@ -56,6 +56,52 @@ impl FactoryView for adw::TabView {
...
+    }
+
+    fn factory_insert_after(
+        &self,
+        widget: impl AsRef<Self::Children>,
+        _position: &(),
+        _other: &Self::ReturnedWidget,
+    ) -> Self::ReturnedWidget {
+        self.add(widget.as_ref());
+        widget.as_ref().clone()
```
Comment (AaronErhardt) [inline relm4/src/factory/widgets/libadwaita.rs:93]: Btw. you can also re-use the `prepend()` method here.

**Comment by tronta** [comment]
Hi @AaronErhardt, what do you think is the best way to go forward here in order to merge it?
In parallel I'm writing a small app which uses this feature where I can test it… with the upcoming Advent of Code and other stuff I'm probably pretty busy :smile: 

Do you think it is ready for review now, or do you think I should wait till I have tested it with an real implementation?

**Comment by AaronErhardt** [comment]
I'm pretty confident that your implementation is correct. And we're still in a beta state officially, so even if there were some minor concerns I wouldn't be afraid of merging this.

File: CHANGES.md [hunk 4]
```diff
@@ -275,6 +275,7 @@
 + core: Pass model in connect_components function of the Widgets trait
 + core: Mini rework of factories
 + core: Removed DefaultWidgets trait in favor of Default implementations in gkt4-rs
++ core. Support factories with adw::PreferencePage
```
Comment (AaronErhardt) [inline CHANGES.md:278]: Can you move this to "Unreleased"?

**Review by AaronErhardt — CHANGES_REQUESTED** [review]

**Review by AaronErhardt — APPROVED** [review]
Thanks!

**Pull request merged**